codec/unknownproto/unknown_fields.go

```diff
@@ -0,0 +1,359 @@
+package unknownproto
+
+import (
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"reflect"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"google.golang.org/protobuf/encoding/protowire"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+)
+
+const bit11NonCritical = 1 << 10
+
+type descriptorIface interface {
+	Descriptor() ([]byte, []int)
+}
+
+// CheckMismatchedFields walks through the protobuf serialized bytes in b, and tries to
+// compare field numbers and wireTypes against what msg expects. The error returned if non-nil will contain
+// the listing of the unknown fields by tagNumber and wireType, or mismatched wireTypes.
+func CheckMismatchedFields(b []byte, msg proto.Message) error {
+	if len(b) == 0 {
+		return nil
+	}
+
+	desc, ok := msg.(descriptorIface)
+	if !ok {
+		return fmt.Errorf("%T does not have a Descriptor() method", msg)
+	}
+
+	fieldDescProtoFromTagNum, _, err := getDescriptorInfo(desc, msg)
+	if err != nil {
+		return err
+	}
+
+	for len(b) > 0 {
+		tagNum, wireType, n := protowire.ConsumeField(b)
+		if n < 0 {
+			return errors.New("invalid length")
+		}
+
+		fieldDescProto, ok := fieldDescProtoFromTagNum[int32(tagNum)]
+		switch {
+		case ok:
+			// Assert that the wireTypes match.
+			if !canEncodeType(wireType, fieldDescProto.GetType()) {
+				return &errMismatchedWireType{
+					Type:         reflect.ValueOf(msg).Type().String(),
+					TagNum:       tagNum,
+					GotWireType:  wireType,
+					WantWireType: protowire.Type(fieldDescProto.WireType()),
+				}
+			}
+
+		default:
+			if tagNum&bit11NonCritical == 0 {
+				// The tag is critical, so report it.
+				return &errUnknownField{
+					Type:     reflect.ValueOf(msg).Type().String(),
+					TagNum:   tagNum,
+					WireType: wireType,
+				}
+			}
+		}
+
+		// Skip over the 2 bytes that store fieldNumber and wireType bytes.
+		fieldBytes := b[2:n]
+		b = b[n:]
+
+		// An unknown but non-critical field or just a scalar type (aka *INT and BYTES like).
+		if fieldDescProto == nil || fieldDescProto.IsScalar() {
+			continue
+		}
+
+		protoMessageName := fieldDescProto.GetTypeName()
+		if protoMessageName == "" {
+			switch typ := fieldDescProto.GetType(); typ {
+			case descriptor.FieldDescriptorProto_TYPE_STRING, descriptor.FieldDescriptorProto_TYPE_BYTES:
+				// At this point only TYPE_STRING is expected to be unregistered, since FieldDescriptorProto.IsScalar() returns false for
+				// TYPE_BYTES and TYPE_STRING as per
+				// https://github.com/gogo/protobuf/blob/5628607bb4c51c3157aacc3a50f0ab707582b805/protoc-gen-gogo/descriptor/descriptor.go#L95-L118
+			default:
+				return fmt.Errorf("failed to get typename for message of type %v, can only be TYPE_STRING or TYPE_BYTES", typ)
+			}
+			continue
+		}
+
+		// Let's recursively traverse and typecheck the field.
+
+		if protoMessageName == ".google.protobuf.Any" {
```

Comment:
We also need to check the `Any` message itself for unknown fields and test that case.

Reply:
That definitely skipped my mind, thank you for the sharp eye!

Reply:
Done, thanks!